File: src/exercise_service_cpp/src/robot_mode_client.cpp
#include "base_interfaces/srv/set_robot_mode.hpp"
#include "rclcpp/rclcpp.hpp"

using base_interfaces::srv::SetRobotMode;
using namespace std::chrono_literals;

class SetRobotModeClient : public rclcpp::Node {
public:
    SetRobotModeClient() : Node("set_robot_mode_client_cpp") {
        RCLCPP_INFO(this->get_logger(), "客户端创建成功!");
        client_ = this->create_client<SetRobotMode>("set_robot_mode");
    }

    bool connect_server() {
        while (!client_->wait_for_service(1s)) {
            if (!rclcpp::ok()) {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "强制退出等待!");
                return false;
            }
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "服务器连接中...");
        }
        return true;
    }

    rclcpp::Client<SetRobotMode>::FutureAndRequestId send_request(std::string mode, float max_speed) {
        auto request = std::make_shared<SetRobotMode::Request>();
        request->mode = mode;
        request->max_speed = max_speed;

        /*
            rclcpp::Client<base_interfaces::srv::SetRobotMode>::FutureAndRequestId
            async_send_request(std::shared_ptr<base_interfaces::srv::SetRobotMode_Request> request)
        */
        return client_->async_send_request(request);
    }

private:
    rclcpp::Client<SetRobotMode>::SharedPtr client_;
};

int main(int argc, char **argv) {
    if (argc != 3) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "提交参数数量不合法!");
        return 1;
    }

    rclcpp::init(argc, argv);
    auto client = std::make_shared<SetRobotModeClient>();

    if (client->connect_server()) {
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "服务器连接成功!");
    } else {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "服务器连接失败!");
        rclcpp::shutdown();
        return 2;
    }

    auto future = client->send_request(argv[1], std::stof(argv[2]));
    if (rclcpp::spin_until_future_complete(client, future) == rclcpp::FutureReturnCode::SUCCESS) {
        auto response = future.get();

        RCLCPP_INFO(client->get_logger(), "请求处理正常!");
        RCLCPP_INFO(client->get_logger(), "====================");
        RCLCPP_INFO(client->get_logger(), "success = %s", response->success ? "true" : "false");
        RCLCPP_INFO(client->get_logger(), "current_mode = %s", response->current_mode.c_str());
        RCLCPP_INFO(client->get_logger(), "current_max_speed = %0.1f", response->current_max_speed);
        RCLCPP_INFO(client->get_logger(), "message = %s", response->message.c_str());
        RCLCPP_INFO(client->get_logger(), "====================");
    } else {
        RCLCPP_ERROR(client->get_logger(), "请求处理异常!");
    }

    rclcpp::shutdown();

    return 0;
}

File: src/exercise_service_py/exercise_service_py/robot_mode_client.py
import sys
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
from base_interfaces.srv import SetRobotMode

class SetRobotModeClient(Node):
    def __init__(self):
        super().__init__("set_robot_mode_client_py")
        self.get_logger().info("客户端创建成功!(Python)")
        self.client_ = self.create_client(SetRobotMode, "set_robot_mode")
        while not self.client_.wait_for_service(1.0):
            get_logger("rclpy").info("服务器连接中...")

    def send_request(self):
        request = SetRobotMode.Request()
        request.mode = sys.argv[1]
        request.max_speed = float(sys.argv[2])
        return self.client_.call_async(request)

def main():
    if len(sys.argv) != 3:
        get_logger("rclpy").error("提交参数数目不合法!")
        return

    rclpy.init()

    client = SetRobotModeClient()

    future = client.send_request()
    rclpy.spin_until_future_complete(client, future)

    try:
        response = future.result()
    except Exception as e:
        client.get_logger().error(f"请求处理异常: {e}")
    else:
        client.get_logger().info("请求正常处理")
        client.get_logger().info("====================")
        client.get_logger().info(f"success = {response.success}")
        client.get_logger().info(f"current_mode = {response.current_mode}")
        client.get_logger().info(f"current_max_speed = {response.current_max_speed}")
        client.get_logger().info(f"message = {response.message}")
        client.get_logger().info("====================")

    rclpy.shutdown()

if __name__ == "__main__":
    main()